### PR TITLE
Harden dhcp by checking for /sbin/dhclient

### PR DESCRIPTION
**Pull request by kris7t**
In order to avoid picking up some random executable from `$PATH` and potentially running it root privileges, `dhcp.c` should check for `/sbin/dhclient` or `/usr/sbin/dhclient` directly. This might cause problems on distributions where `dhclient` is located elsewhere, but as far as I know, `/sbin` is the default location (and even Arch, which uses only `/usr/bin`, symlinks `/sbin` to `/usr/bin`). This is the same approach that `netfilter.c` takes when running `iptables-restore`.

I formulated the check for root privileges according to https://github.com/netblue30/firejail/blob/master/src/firejail/x11.c#L1103, which allows the binary to either be owned by root, or the root group. A more paranoid version would use `open`, `fstat` and `fexecve` to make sure ownership of the binary does not change between the `stat` and `exec` calls. On the other hand, manipulation of a binary already owned by root means the system is already compromised, so there's probably not much point to be careful about TOCTOU here. `netfiler.c` even skips the check altogether, and relies on `/sbin` being only writable by root.

Hopefully, we don't blindly execute a binary with root privileges anywhere else in the codebase.

_A more meta question: what is the policy with running external executable from firejail with root privilages (even if capability and seccomp sandboxed)?_ The potential for introducing setuid-related vulnerabilities is high. If we want to integrate additional external tools (like `xdg-dbus-proxy`), we should try to keep it reasonably safe.

### CODE REVIEW

File: src/firejail/dhcp.c [hunk 1]
```diff
@@ -100,7 +101,7 @@ static pid_t dhcp_read_pidfile(const Dhclient *client) {
 			long pid;
 			if (fscanf(pidfile, "%ld", &pid) == 1) {
 				char *pidname = pid_proc_comm((pid_t) pid);
-				if (pidname && strcmp(pidname, "dhclient") == 0)
+				if (pidname && strcmp(pidname, dhclient_path))
```
Comment (smitsohu) [inline src/firejail/dhcp.c]: you don't want equal strings any more?

File: src/firejail/dhcp.c [hunk 2]
```diff
@@ -139,18 +140,32 @@ void dhcp_start(void) {
 	if (!any_dhcp())
 		return;
 
+	char *dhclient_path = "/sbin/dhclient";
+	struct stat s;
+	if (stat(dhclient_path, &s) == -1) {
+		dhclient_path = "/usr/sbin/dhclient";
+		if (stat(dhclient_path, &s) == -1) {
+			fprintf(stderr, "Error: dhclient was not found.\n");
+			exit(-1);
```
Comment (smitsohu) [inline src/firejail/dhcp.c]: Can you return `1`? Just to keep it consistent.

**Comment by smitsohu** [comment]
IMHO sbox is not really in the position to figure out what to execute. Maybe we should replace the `execvp` with `execv`?

File: src/firejail/dhcp.c (hunk 1 above)
Comment (kris7t) [reply src/firejail/dhcp.c]: Whoops, good catch!

Comment (kris7t) [reply src/firejail/dhcp.c]: I don't know why I was fumbling around with this. `pid_proc_comm` returns `"dhclient"` even if it was started as `/sbin/dhclient`, so the previous check was actually correct.

**Comment by kris7t** [comment]
Yeah, `execv` sounds much safer. We could even move the uid check to `sbox`, since it would receive the absolute path of the executable.

File: src/firejail/dhcp.c (hunk 2 above)
Comment (kris7t) [reply src/firejail/dhcp.c]: Done.

**Comment by smitsohu** [comment]
> Hopefully, we don't blindly execute a binary with root privileges anywhere else in the codebase.

Quickly checked and there seem to be no other instances. Also means that replacing `execvp` with `execv` is indeed a trivial thing.

> We could even move the uid check to sbox, since it would receive the absolute path of the executable.

:+1:

**Comment by smitsohu** [comment]
Merged, thanks!

**Comment by smitsohu** [comment]
I'm going to remove the execvp in a separate patch.

**Comment by kris7t** [comment]
@smitsohu  Oh, I am also trying to put together a patch for -- with some extra checking. Playing around with some symlinks, I think the paranoia with `fecexve` might be warranted, although I am still figuring out how to use it properly.

**Comment by smitsohu** [comment]
@kris7t Go ahead! I did not start yet :)

**Pull request merged**